hazelcast/src/main/java/com/hazelcast/query/impl/SortedIndexStore.java

```diff
@@ -178,10 +167,109 @@ public void clear() {
         }
     }
 
+    /**
+     * Adds entry to the given index map without copying it.
+     *
+     * @see IndexCopyBehavior
+     */
+    private class AddFunctor implements IndexFunctor<Comparable, QueryableEntry> {
+        @Override
+        public void invoke(Comparable attribute, QueryableEntry entry) {
+            if (attribute instanceof IndexImpl.NullObject) {
+                recordsWithNullValue.put(entry.getKeyData(), entry);
+            } else {
+                Map<Data, QueryableEntry> records = recordMap.get(attribute);
+                if (records == null) {
+                    records = new ConcurrentHashMap<Data, QueryableEntry>(1, LOAD_FACTOR, 1);
+                    recordMap.put(attribute, records);
+                }
+                records.put(entry.getKeyData(), entry);
+            }
+        }
+    }
+
+    /**
+     * Adds entry to the given index map copying it to secure exclusive access.
+     *
+     * @see IndexCopyBehavior
+     */
+    private class CopyOnWriteAddFunctor implements IndexFunctor<Comparable, QueryableEntry> {
```

Comment:
can we comment that `CopyOnWriteAddFunctor` must be accessed in a thread safe manner?